### ISSUE TITLE
Why does the order of HTTP parameters matter?

### ISSUE THREAD

**Issue by shadyxu** [OpenTSDB/opentsdb#411]
Why would
`curl http://localhost:4242/api/suggest?type=metrics&q=sys`
returns the expected result, but
`http://localhost:4242/api/suggest?q=sys&type=metrics`
returns a `Missing Parameter` error?

This is definitely not as expected, is it?

**Comment by oozie** [comment]
I think we need to zoom out for a moment.
If you type `curl http://localhost:4242/api/suggest?type=metrics&q=sys` in bash, the & sign puts curl in the background, and the command that's effectively executed is `curl http://localhost:4242/api/suggest?type=metrics`. As pointed out by @pramitchoudhary, this is avalid query returning the top bunch of metrics. Let's switch it up now:
```
oozie@number5:~$ curl http://localhost:4242/api/suggest?q=sys&type=metrics
[1] 3371
[...] missing parameter [...]
[1]+  Done                    curl http://localhost:4242/api/suggest?q=sys
```
And that query is missing a parameter ^^^^
Once you quote the request, it works as expected:
```
oozie@number5:~$ curl "http://localhost:4242/api/suggest?q=sys&type=metrics"
["sys.numa.allocation","sys.numa.foreign_allocs","sys.numa.interleave","sys.numa.zoneallocs"]
oozie@number5:~$ curl "http://localhost:4242/api/suggest?type=metrics&q=sys"
["sys.numa.allocation","sys.numa.foreign_allocs","sys.numa.interleave","sys.numa.zoneallocs"]
```
Does this crack the case?

**Comment by shadyxu** [comment]
@oozie Yes you are right. But I use `curl` here just to simplify the question. Actually I am using opentsdb with Grafana which does not have the problem you mentioned.

**Comment by oozie** [comment]
I also used `curl` to illustrate the point - once the string is properly quoted, both queries return the exact same result. I can't think of a reason why that wouldn't be the case for other clients. It is concievable, however, that Grafana executes a shell command to get the datapoints and falls for the same trap. OpenTSDB logs should also contain definitive hints.
If the problem persists, how can it be reproduced?

**Comment by shadyxu** [comment]
@oozie The problem and solution to it you mentioned works only in shell environment. Grafana does not use a shell command to fetch datapoints, however, it sends HTTP request.

**Comment by oozie** [comment]
Clearly, I haven't used Grafana before. I've just set it up now to repro, and everything works as expected, even when the `q=` parameter comes before `type=`.
![grafana-opentsdb](https://cloud.githubusercontent.com/assets/1587140/6462448/24bfdcb6-c15e-11e4-925f-6c30e4221f49.png)

I'm still inclined to believe that there never was a problem with Netty or OpenTSDB with regards to server-side query string parsing. Such behavior would be a serious problem for other queries and it would manifest itself early.

The only logical explanation for `Missing Parameter` problem is that the parameter never made it into the query, like in a curl example above.
